atomic_reactor/plugins/exit_delete_from_registry.py

```diff
@@ -0,0 +1,79 @@
+All rights reserved.
+
+This software may be modified and distributed under the terms
+of the BSD license. See the LICENSE file for details.
+"""
+
+from __future__ import unicode_literals
+
+from copy import deepcopy
+import requests
+import requests.auth
+try:
+    from urlparse import urlparse
+except ImportError:
+    from urllib.parse import urlparse
+
+from atomic_reactor.plugin import ExitPlugin, PluginFailedException
+from atomic_reactor.util import Dockercfg
+
+class DeleteFromRegistryPlugin(ExitPlugin):
+    """
+    Delete previously pushed v2 images from a registry.
+    """
+
+    key = "delete_from_registry"
+    is_allowed_to_fail = False
+
+    def __init__(self, tasker, workflow, registries):
+        """
+        :param tasker: DockerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        :param registries: dict, keys are docker registries, values are dicts containing per-registry
+                           parameters.
+                           Params:
+                            * "secret" optional string - path to the secret, which stores
+                              login and password for remote registry
+        """
+        super(DeleteFromRegistryPlugin, self).__init__(tasker, workflow)
+
+        self.registries = deepcopy(registries)
+
+    def run(self):
+        for registry, registry_conf in self.registries.items():
+            if not registry.startswith('http://') and not registry.startswith('https://'):
+                registry = 'https://' + registry
+
+            registry_noschema = urlparse(registry).netloc
+
+            auth = None
+            secret_path = registry_conf.get('secret')
+            if secret_path:
+                dockercfg = Dockercfg(secret_path)
+                username = dockercfg.get_credentials(registry_noschema)['username']
+                password = dockercfg.get_credentials(registry_noschema)['password']
+                auth = requests.auth.HTTPBasicAuth(username, password)
+
+            for push_conf_registry in self.workflow.push_conf.docker_registries:
+                if push_conf_registry.uri == registry_noschema:
+                    break
+            else:
+                self.log.warning("requested deleting image from %s but we haven't pushed there", registry_noschema)
+                continue
+
+            for tag, digest in push_conf_registry.digests.items():
+                repo = tag.split(':')[0]
+                url = registry + "/v2/" + repo + "/manifests/" + digest
+                insecure = push_conf_registry.insecure
+                response = requests.delete(url, verify=not insecure, auth=auth)
+
+                if response.status_code == 202:
+                    self.log.info("deleted manifest %s/%s@%s", registry_noschema, repo, digest)
+                elif response.status_code == 404:
+                    self.log.warning("cannot delete %s/%s@%s: not found", registry_noschema, repo, digest)
```

Comment:
We should also handle requests.codes.method_not_allowed with a warning not a failure.